x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java

```diff
@@ -429,11 +437,19 @@ public static void buildRoleFromDescriptors(
                     }
                 });
             }
+
+            if (descriptor.hasWorkflowsRestriction()) {
+                if (roleDescriptors.size() != 1) {
+                    throw new IllegalArgumentException("only single role descriptor with workflow restriction is allowed");
+                }
```

Comment:
I suggest we check this in the beginning of the method (right after checking whether list of role descriptor is empty) and call `listener.onFailure` if there is violation (instead of throwing). Also nit: can we add the role descriptor names in the error message?

Reply:
Thanks for the suggestion. I've implemented the [validation](https://github.com/elastic/elasticsearch/pull/96744/files#diff-65dc248b6cd87fca037bd9e993b2aab3828578878ab92e507a88ce89925b493eR385-R408) and [added tests](https://github.com/elastic/elasticsearch/pull/96744/files#diff-57010f0ef921acba217dce681c900c49393738fd80e79f0709cb2a241ea1a558R1093-R1118) for it.